### PR TITLE
Offline: UI improvements

### PR DESCRIPTION
**Pull request by des12437**
Related to `swift/6048`

- Rephrases empty state view text
- Uses `ContentUnavailableView` when map is offline disabled
- Removes HTML tags from map area description
- Animates open button

### CODE REVIEW

File: Sources/ArcGISToolkit/Components/Offline/OfflineMapAreasView.swift [hunk 1]
```diff
@@ -102,22 +110,33 @@ public struct OfflineMapAreasView: View {
...
-            Text("You don't have any map areas yet.")
+            Text("There are no map areas defined for this web map.")
                 .font(.subheadline)
                 .foregroundStyle(.secondary)
+                .multilineTextAlignment(.center)
         }
         .frame(maxWidth: .infinity)
     }
     
     private var offlineDisabledView: some View {
```
Comment (philium) [inline Sources/ArcGISToolkit/Components/Offline/OfflineMapAreasView.swift:121]: Please remove the explicit return statements and then annotate with `ViewBuilder`. Also, consolidate the strings.

```swift
@ViewBuilder private var offlineDisabledView: some View {
    let labelText = Text("Offline Disabled")
    let descriptionText = Text("Please ensure the web map is offline enabled.")
    if #available(iOS 17, *) {
        ContentUnavailableView {
            labelText
                .bold()
        } description: {
            descriptionText
        }
    } else {
        VStack(alignment: .center) {
            labelText
                .bold()
                .font(.title2)
            descriptionText
                .font(.subheadline)
                .foregroundStyle(.secondary)
        }
        .frame(maxWidth: .infinity)
    }
}
```

File: Sources/ArcGISToolkit/Components/Offline/PreplannedMapModel.swift [hunk 2]
```diff
@@ -318,7 +320,8 @@ extension PreplannedMapArea: PreplannedMapAreaProtocol {
     }
     
     var description: String {
-        portalItem.description
+        // Remove HTML tags from description.
+        portalItem.description.replacingOccurrences(of: "<[^>]+>", with: "", options: .regularExpression)
```
Comment (philium) [inline Sources/ArcGISToolkit/Components/Offline/PreplannedMapModel.swift:324]: Please use a Swift regex expression instead.

```suggestion
        portalItem.description.replacing(/<[^>]+>/, with: "")
```

Note: This suggestion will requiring enabling the `BareSlashRegexLiterals` upcoming feature:

```swift
.enableUpcomingFeature("BareSlashRegexLiterals")
```

Which can then be removed as part of the work to move to Swift 6 as that feature is enabled by default.

File: Tests/ArcGISToolkitTests/PreplannedMapModelTests.swift [hunk 3]
```diff
@@ -374,6 +374,34 @@ class PreplannedMapModelTests: XCTestCase {
...
+            // User notifications in unit tests are not supported, must pass false here
+            // or the test process will crash.
+            showsUserNotificationOnCompletion: false
+        )
+        
+        // Verify description does not contain HTML tags.
+        XCTAssertEqual(model.preplannedMapArea.description,
+        """
+        A map that contains stormwater network within Naperville, IL, USA with cartography designed for web and mobile devices. This is a demo map to demonstrate offline capabilities with ArcGIS Runtime and is based on ArcGIS Solutions for Stormwater.
+        """)
```
Comment (philium) [inline Tests/ArcGISToolkitTests/PreplannedMapModelTests.swift:403]: This string literal only has a single line, so a multiline string literal shouldn't be used. Also, here is how we would format/indent this assertion:

```suggestion
        XCTAssertEqual(
            model.preplannedMapArea.description,
            "A map that contains stormwater network within Naperville, IL, USA with cartography designed for web and mobile devices. This is a demo map to demonstrate offline capabilities with ArcGIS Runtime and is based on ArcGIS Solutions for Stormwater."
        )
```

**Review by yo1995 — COMMENTED** [review]
LGTM with Phil suggested changes

File: Sources/ArcGISToolkit/Components/Offline/PreplannedMapModel.swift [hunk 4]
```diff
@@ -130,7 +130,9 @@ class PreplannedMapModel: ObservableObject, Identifiable {
     func updateDownloadStatus(for downloadResult: Result<DownloadPreplannedOfflineMapResult, any Error>) {
         switch downloadResult {
         case .success:
-            status = .downloaded
+            withAnimation(.easeInOut) {
```
Comment (philium) [inline Sources/ArcGISToolkit/Components/Offline/PreplannedMapModel.swift:133]: This really shouldn't be done by the model as `withAnimation` is a SwiftUI function. I just noticed that this file imports SwiftUI, which is a huge code smell. The model should know nothing about the view; the view itself should perform the animation in response to a change in the model's state.

Comment (des12437) [reply Sources/ArcGISToolkit/Components/Offline/PreplannedMapModel.swift:133]: Yes using `withAnimation` in the model was a workaround. The `animation<V>(_ animation: Animation?, value: V)` view modifier requires a value to change to animate, and `model.status` is not equatable. I tried using `model.status.isDownloaded` but that does not trigger an animation. Do you suggest making `model.status` equatable or another option?

Comment (philium) [reply Sources/ArcGISToolkit/Components/Offline/PreplannedMapModel.swift:133]: There is another option: `View.onReceive(_:perform:)`. First, revert these changes and change `import SwiftUI` to `import Combine`. Then, in the list item view, add a new download state type and state property:

```swift
private enum DownloadState {
    case notDownloaded, downloading, downloaded
    
    init(_ state: PreplannedMapModel.Status) {
        self = switch state {
        case .downloaded: .downloaded
        case .downloading: .downloading
        default: .notDownloaded
        }
    }
}

@State private var downloadState: DownloadState = .notDownloaded
```

Then, add a couple modifiers to set the initial download state and update the download state in response to the model status changes:

```swift
.onAppear {
    downloadState = .init(model.status)
}
.onReceive(model.$status) { status in
    let downloadState = DownloadState(status)
    withAnimation(
        downloadState == .downloaded ? .easeInOut : nil
    ) {
        self.downloadState = downloadState
    }
}
```

Lastly, update the code in `downloadButton` to use the new download state.

File: Sources/ArcGISToolkit/Components/Offline/PreplannedListItemView.swift [hunk 5]
```diff
@@ -27,6 +27,22 @@ struct PreplannedListItemView: View {
...
+    private enum DownloadState {
+        case notDownloaded, downloading, downloaded
+        
+        init(_ state: PreplannedMapModel.Status) {
+            self = switch state {
+            case .downloaded: .downloaded
+            case .downloading: .downloading
+            default: .notDownloaded
+            }
+        }
```
Comment (philium) [inline Sources/ArcGISToolkit/Components/Offline/PreplannedListItemView.swift:40]: I just showed this here for simplicity, but normally we would keep an enum/struct declaration clean of convenience members and instead add them in private extensions towards the bottom of the file. Consider doing that here to keep the view itself as simple and readable as possible.

**Review by philium — APPROVED** [review]

File: Sources/ArcGISToolkit/Components/Offline/PreplannedListItemView.swift (hunk 5 above)
Comment (philium) [reply Sources/ArcGISToolkit/Components/Offline/PreplannedListItemView.swift]: You misunderstood what I meant. I was referring to this initializer (the one that I selected) only and not the entire enum. The enum itself should be where it was, but this initializer makes sense in an extension:

```swift
private extension PreplannedListItemView.DownloadState {
    init(_ state: PreplannedMapModel.Status) {
        ...
    }
}
```

File: Sources/ArcGISToolkit/Components/Offline/PreplannedListItemView.swift [hunk 6]
```diff
@@ -27,6 +27,14 @@ struct PreplannedListItemView: View {
     /// A Boolean value indicating whether the metadata view is presented.
     @State private var metadataViewIsPresented = false
     
+    /// The download state of the preplanned map model.
+    enum DownloadState {
```
Comment (philium) [inline Sources/ArcGISToolkit/Components/Offline/PreplannedListItemView.swift:31]: This should still be private.

```suggestion
    fileprivate enum DownloadState {
```

**Review by philium — APPROVED** [review]

File: Sources/ArcGISToolkit/Components/Offline/PreplannedMapModel.swift [hunk 7]
```diff
@@ -14,7 +14,7 @@
 
 import ArcGIS
 import OSLog
-import SwiftUI
+import Combine
```
Comment (yo1995) [inline Sources/ArcGISToolkit/Components/Offline/PreplannedMapModel.swift:17]: When I comment out this import, it still compiles.

```suggestion
```

It seems the import of `OSLog` implicitly introduced `Foundation`. I'm not sure if we should import Foundation here as well.

**Review by yo1995 — APPROVED** [review]

File: Sources/ArcGISToolkit/Components/Offline/PreplannedMapModel.swift (hunk 7 above)
Comment (philium) [reply Sources/ArcGISToolkit/Components/Offline/PreplannedMapModel.swift:17]: Oh! Good catch. `OSLog` should not be imported; it has been superseded by `os` (which does not export Foundation, by the way). So the imports for this file should be as follows:

```swift
import ArcGIS
import Combine
import Foundation

internal import os
```

**Review by philium — APPROVED** [review]

**Review by yo1995 — APPROVED** [review]

**Comment by des12437** [comment]
Thank you both!

**Pull request merged**